### PR TITLE
fix(issue): mcp_discover hangs when parallel stdio connections open concurrent trust dialogs

### LINKED ISSUE #1449: mcp_discover hangs when parallel stdio connections open concurrent trust dialogs
## Problem

When an assistant issues first-time `mcp_discover` calls for two unapproved stdio MCP servers in the same parallel tool group, the interactive session can hang indefinitely.

This happened twice on GSD 1.11.0:

- occurrence 1 remained pending for at least 18 minutes 3 seconds;
- occurrence 2 remained pending for at least 54 minutes 51 seconds;
- neither `mcp_discover` call produced a matching tool result;
- both sessions required a restart.

The MCP servers themselves were ruled out: isolated SDK probes completed `initialize` and `tools/list` in milliseconds. No child process corresponding to either failed discovery was started, placing the stall before stdio transport creation.

## Root cause

`src/resources/extensions/mcp-client/index.ts:82-111` opens a blocking `ctx.ui.confirm` for every unapproved stdio server:

```ts
const approved = await ctx.ui.confirm(
  `Trust MCP server "${config.name}"?`,
  `Project config ${config.sourcePath} wants to start: ...`,
  _buildMcpTrustConfirmOptionsForTest(signal),
);
```

The extension UI documentation describes `ctx.ui.confirm` as blocking and temporarily replacing the editor.

`getOrConnect()` deduplicates pending connections only by server name (`index.ts:125-143`):

```ts
const pending = pendingConnections.get(config.name);
if (pending) return pending;

const connectionPromise = connectServer(config, signal, ctx);
pendingConnections.set(config.name, connectionPromise);
```

This handles duplicate discovery of one server, but two different servers may enter separate blocking confirmation dialogs concurrently. There is no prompt mutex or queue in the MCP extension.

The transport is created only after approval (`index.ts:146-159`):

```ts
if (config.transport === "stdio" && config.command) {
  approvedTrustKey = await assertTrustedStdioServer(config, ctx, signal);
  transport = new StdioClientTransport(...);
}
```

Consequently, the 30-second `client.connect()` and `listTools()` timeouts do not cover a call stalled in the earlier approval phase.

Although the confirmation receives `{ timeout: 120_000, signal }`, the observed tool calls remained pending far beyond 120 seconds.

## Expected behavior

First-time stdio approvals must not orphan or indefinitely block one another.

Concurrent discoveries should either:

1. queue blocking trust dialogs and complete both calls in order; or
2. reject the second call promptly with an actionable error.

Every approval wait must have a hard abortable timeout that rejects the tool call and releases any queue state.

## Suggested fix

Serialize ownership of the blocking trust dialog with a module-level mutex or promise queue. Only the approval prompt needs serialization; MCP transport startup and discovery can remain parallel after approval.

After waiting for the queue, re-check:

- the caller's `AbortSignal`;
- whether the server was approved while waiting;
- whether the session is shutting down.

Wrap the full approval phase in an extension-owned hard timeout rather than relying only on UI option semantics. Release the queue in `finally`.

A timeout should return a phase-specific error such as:

```text
Timed out waiting for stdio MCP trust approval for "<server>"
```

The core extension UI could also be hardened so concurrent blocking `ctx.ui.select/confirm/input/editor` calls are centrally serialized or explicitly rejected rather than allowing ambiguous dialog ownership.

## Missing regression coverage

`src/resources/extensions/gsd/tests/mcp-client-security.test.ts` currently verifies only that confirmation options contain the expected timeout and signal:

```ts
const options = _buildMcpTrustConfirmOptionsForTest(controller.signal);

assert.equal(options.timeout, 120_000);
assert.equal(options.signal, controller.signal);
```

This does not test concurrent dialogs or bounded completion.

Please add behavioral tests that:

1. register two fake, unapproved stdio MCP configurations;
2. invoke their first `mcp_discover` calls concurrently;
3. assert that only one confirmation callback is active at a time;
4. resolve the first prompt and assert that the second appears;
5. resolve the second and assert that both discoveries complete;
6. abort a discovery while it waits and assert prompt rejection and queue cleanup;
7. leave a prompt unresolved and assert rejection at the configured timeout;
8. invoke two discoveries for the same server and confirm that `pendingConnections` still produces one prompt and one connection.

## Reproduction

1. Configure two stdio MCP servers that have not yet been approved in the current session.
2. Issue two `mcp_discover` calls in the same parallel assistant tool group.
3. Observe that the tool turn remains pending and neither discovery returns.

## Environment

- GSD: 1.11.0
- Mode: interactive session
- Tool: `mcp_discover`
- Transport: stdio

## Forensic evidence

- Two independent occurrences involved the same pair of parallel first-time stdio discoveries.
- Neither occurrence recorded matching tool results.
- Minimum observed stalls were 18m03s and 54m51s.
- Isolated MCP SDK probes completed handshake and `tools/list` in milliseconds.
- No child process corresponding to the failed discoveries was started.
- No auto-mode lock, paused session, active milestone, or active slice was involved.
- Historical `artifact_file_missing` doctor findings were investigated and ruled out as unrelated.

---

*Auto-generated by `/gsd forensics`*

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary
- Serialized stdio MCP trust approvals with hard timeout and abort cleanup, added focused regression coverage, and verified syntax/diff checks while full tests were blocked by missing dependencies and network.

## Verification
- Completed in the repository worktree before push.

## Related Issue
- Closes #1449
- [#1449 mcp_discover hangs when parallel stdio connections open concurrent trust dialogs](https://github.com/open-gsd/gsd-pi/issues/1449)

## User
- @rager306

## Repo
- `open-gsd/gsd-pi`

## Branch
- `issue/1449-mcp-discover-hangs-when-parallel-stdio-c-1784416081`

---

> [!NOTE]
> **Medium Risk**
> Changes interactive trust and connection gating for project-local stdio MCP servers; mistakes could block approvals or leave waiters stuck, but scope is limited to the MCP client extension with substantial regression tests.
> 
> **Overview**
> Fixes hangs when parallel stdio MCP connections tried to open **multiple trust dialogs at once** (e.g. concurrent `mcp_discover`).
> 
> Stdio trust approval now runs through a **serialized prompt queue** so only one `ui.confirm` runs at a time. Each waiter gets a **120s hard timeout** that starts when it actually reaches the prompt (not while queued), plus **caller abort** support for both active and queued waiters. **`closeAll`** (session shutdown/switch) rejects active and queued trust waiters with an MCP-session-closed error and resets the queue.
> 
> Adds **behavior-focused tests** (replacing removed grep-style checks): queue ordering, abort/timeout edge cases, shutdown cleanup, and an integration test that concurrent `mcp_discover` for the same server shows one trust prompt and one stdio server start.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d7efde8c18baea1a31c26e54bcf803f3db3e29f4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/open-gsd/codesmith/gsd-pi/pr/1496"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is enabled.</sup>

### CODE REVIEW

**Comment by jeremymcs** [comment]
✅ **Verification — all 1 task is addressed**

Re-checked PR #1496 against [#1449 mcp_discover hangs when parallel stdio connections open concurrent trust dialogs](https://github.com/open-gsd/gsd-pi/issues/1449).

## Subtasks
- [x] **mcp_discover hangs when parallel stdio connections open concurrent trust dialogs**

_Re-run from Patchdeck to refresh._

File: src/resources/extensions/mcp-client/index.ts [hunk 1]
```diff
@@ -75,14 +79,102 @@ export function _buildMcpChildEnvForTest(configEnv: Record<string, string> | und
...
+
+	const approvalAbort = createTrustApprovalAbort(serverName, signal, timeout);
+	try {
+		await Promise.race([previous.catch(() => {}), approvalAbort.aborted]);
+		if (approvalAbort.signal.aborted) await approvalAbort.aborted;
+		return await Promise.race([run(approvalAbort.signal), approvalAbort.aborted]);
+	} finally {
+		approvalAbort.cleanup();
+		releasePrompt();
+		if (stdioTrustPromptQueue === queued) stdioTrustPromptQueue = emptyStdioTrustPromptQueue;
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d7efde8`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d7efde8`.

File: src/resources/extensions/mcp-client/index.ts [hunk 2]
```diff
@@ -75,14 +79,101 @@ export function _buildMcpChildEnvForTest(configEnv: Record<string, string> | und
...
+		releasePrompt = resolve;
+	});
+	const queued = previous.catch(() => {}).then(() => current);
+	stdioTrustPromptQueue = queued;
+
+	const approvalAbort = createTrustApprovalAbort(serverName, signal, timeout);
+	try {
+		await Promise.race([previous.catch(() => {}), approvalAbort.aborted]);
+		if (approvalAbort.signal.aborted) await approvalAbort.aborted;
+		return await Promise.race([run(approvalAbort.signal), approvalAbort.aborted]);
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d7efde8`.

**Comment by jeremymcs** [comment]
**Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d7efde8`.

File: src/resources/extensions/mcp-client/index.ts [hunk 3]
```diff
@@ -192,6 +325,8 @@ async function connectServer(config: McpServerConfig, signal?: AbortSignal, ctx?
 }
 
 async function closeAll(): Promise<void> {
+	abortActiveStdioTrustApprovals(new Error("MCP session closed while waiting for stdio MCP trust approval."));
+	stdioTrustPromptQueue = emptyStdioTrustPromptQueue;
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts:349]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `d7efde8`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/mcp-client/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts]: Addressed in commit `d7efde8`.

No additional changes were needed; the current implementation no longer resets the global stdio trust prompt queue from an individual queued waiter’s finally path, preserving serialization while another prompt is active.

**Comment by jeremymcs** [comment]
Addressed in commit `d7efde8`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1496#pullrequestreview-4729482855):
> 

Confirmed the queue-reset issue is already fixed in the current worktree and covered by the focused MCP client security regression tests.

File: src/resources/extensions/mcp-client/index.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts]: Addressed in commit `d7efde8`.

No additional changes were needed; `createTrustApprovalAbort` is already deferred until after the serialized queue wait completes, so the trust timeout only covers the active prompt window.

**Comment by jeremymcs** [comment]
Addressed in commit `d7efde8`.

> Responding to [comment by @cursor[bot]](https://github.com/open-gsd/gsd-pi/pull/1496#pullrequestreview-4729485663):
> 

Confirmed the queued-timeout issue is already fixed in the current worktree and covered by the focused MCP client security regression tests.

File: src/resources/extensions/mcp-client/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/mcp-client/index.ts:349]: Addressed in commit `d7efde8`.

No additional changes were needed; `closeAll` already aborts both active and queued stdio trust approvals, preventing queued waiters from reaching `ui.confirm` after shutdown.
